File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        ans = []
        digitsMapped = {
            "2" : "abc",
            "3" : "def",
            "4" : "ghi",
            "5" : "jkl",
            "6" : "mno",
            "7" : "pqrs",
            "8" : "tuv",
            "9" : "wxyz"
        }

        def backtracking(idx, currStr):
            if len(currStr) == len(digits):
                ans.append(currStr)
                return
            for c in digitsMapped[digits[idx]]:
                backtracking(idx + 1, currStr + c)
        
        if digits:
            backtracking(0, "")
        
        return ans

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> ans;
    vector<string> digitsMapped = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    void helperBT(int idx, string currStr, string& digits) {
        if(currStr.length() == digits.length()) {
            ans.push_back(currStr);
            return;
        }
        string chars = digitsMapped[digits[idx] - '0'];
        for(char c : chars)
            helperBT(idx + 1, currStr + c, digits);
    }

    vector<string> letterCombinations(string digits) {
        if(digits.empty())
            return ans;
        helperBT(0, "", digits);
        return ans;
    }
};